pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/security/xml/ApexSOQLInjection.xml

```diff
@@ -322,6 +323,329 @@ public class Foo {
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public void test1() {
+        String field1 = 'someIDhere';
+        String field2 = 'SELECT COUNT() FROM Account WHERE Id =';
+        Database.countQuery(field2 + field1);
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Safe SOQL on concatenation of variables</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public void test1() {
+        String field1 = String.escapeSingleQuotes('yo');
+        String field2 = 'SELECT COUNT() FROM Account WHERE Id =' + field1;
+        Database.countQuery(field2);
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Safe SOQL + merged variable from a literal</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public void test1() {
+        String field1 = String.escapeSingleQuotes('yo');
+        Database.countQuery('SELECT COUNT() FROM Account WHERE Id =' + field1);
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Safe SOQL + merged variable</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+
+    public void test1(String field1) {
+        field2 = String.escapeSingleQuotes(field1);
+        Database.countQuery('SELECT COUNT() FROM Account WHERE Id =' + field2);
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>No issue when SOQL is called</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public void test1() {
+        Database.countQuery('SELECT COUNT() FROM Account');
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Potentially unsafe SOQL with variable concatenation</description>
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public void test1(String t1) {
+        Database.countQuery('SELECT COUNT() FROM Account' + t1);
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Potentially unsafe SOQL with multiple variable concatenation</description>
+        <expected-problems>2</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public void test1(String t1, String t2) {
+        Database.countQuery('SELECT COUNT() FROM Account ' + t1 + ' WHERE ' + t2 + ';');
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Safe SOQL with List concatenation</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public void test1(List<String> t1) {
+        Database.countQuery('SELECT COUNT() FROM Account ' + string.join(t1, ','));
```

Comment:
I know that this is the same test case as for Database.query, but: Wouldn't that allow injection here? (in both cases)

The list t1 is coming from somewhere unsafe. If the first (and only) entry in the list is e.g. "; DELETE FROM Account" - wouldn't that just be executed? Since in this case the list has only 1 element, it is exactly like a simple concatenation...

Would you mind following up on this? If you agree with my assessment, please create a new issue - so that somebody can fix it for both methods (query, countQuery).

I'd merge this PR as is for now, since it is consistent with the current behavior of the rule. This potential bug should be fixed in a separate PR.

Reply:
Yeah, I've managed to inject in that case:

```apex
List<string> stuff = new List<string>();

stuff.add('WHERE Name = \'hi\' OR Name = \'');
stuff.add('\'');

Database.countQuery(
    'SELECT COUNT() FROM Account ' +
    string.join(stuff, ',')
);
```

I'll open an issue and if I have some time in the near future I'll try fixing that one :)